idx/index/types.go

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+import (
+	"github.com/m3db/m3ninx/idx/doc"
+)
+
+// Index is a collection of searchable documents.
+type Index interface {
+	Writer
```

Comment:
does an Index have to be writable? e.g. when TTL'ing in m3db, we'll have multiple immutable indexes, and one mutable index, with whatever types of segments internal to each.

Reply:
I modified this interface slightly to reflect that. It now includes a `Seal` method which will mark the index as immutable. Currently I have the `Seal` method returning an iterator but I could also envision it returning a `Snapshot` reflecting the state of the index after it was closed. Personally I'm leaning towards the latter so that we're explicit about the Index no longer being writable as opposed to expecting the client to maintain state on which index is sealed or not themselves.

Reply:
I don't think Seal() can return an Iterator because you still need to be able to query the resulting type. 

Returning a Snapshot would work as long as we could query it (think you already have the types for this), and iterate the documents in it.

Reply:
Could we just return a segment instead?